graph/src/blockchain/firehose_block_stream.rs

```diff
@@ -0,0 +1,252 @@
+use crate::sf::bstream;
+
+pub struct FirehoseBlockStreamContext<C>
+where
+    C: Blockchain,
+{
+    cursor: Option<String>,
+    mapper: Arc<C::FirehoseMapper>,
+    node_id: NodeId,
+    subgraph_id: DeploymentHash,
+    filter: Arc<C::TriggerFilter>,
+    start_blocks: Vec<BlockNumber>,
+    logger: Logger,
+}
+
+impl<C: Blockchain> Clone for FirehoseBlockStreamContext<C> {
+    fn clone(&self) -> Self {
+        Self {
+            cursor: self.cursor.clone(),
+            mapper: self.mapper.clone(),
+            node_id: self.node_id.clone(),
+            subgraph_id: self.subgraph_id.clone(),
+            filter: self.filter.clone(),
+            start_blocks: self.start_blocks.clone(),
+            logger: self.logger.clone(),
+        }
+    }
+}
+
+enum BlockStreamState {
+    Disconneted,
+    Connecting(
+        Pin<
+            Box<
+                dyn futures03::Future<
+                    Output = Result<tonic::Streaming<bstream::BlockResponseV2>, anyhow::Error>,
+                >,
+            >,
+        >,
+    ),
+    Connected(tonic::Streaming<bstream::BlockResponseV2>),
+}
+
+pub struct FirehoseBlockStream<C: Blockchain> {
+    endpoint: Arc<FirehoseEndpoint>,
+    state: BlockStreamState,
+    ctx: FirehoseBlockStreamContext<C>,
+    connection_attempts: u64,
+}
+
+impl<C> FirehoseBlockStream<C>
+where
+    C: Blockchain,
+{
+    pub fn new(
+        endpoint: Arc<FirehoseEndpoint>,
+        cursor: Option<String>,
+        mapper: Arc<C::FirehoseMapper>,
+        node_id: NodeId,
+        subgraph_id: DeploymentHash,
+        filter: Arc<C::TriggerFilter>,
+        start_blocks: Vec<BlockNumber>,
+        logger: Logger,
+    ) -> Self {
+        FirehoseBlockStream {
+            endpoint,
+            state: BlockStreamState::Disconneted,
+            ctx: FirehoseBlockStreamContext {
+                cursor,
+                mapper,
+                node_id,
+                subgraph_id,
+                logger,
+                filter,
+                start_blocks,
+            },
+            connection_attempts: 0,
+        }
+    }
+}
+
+impl<C: Blockchain> BlockStream<C> for FirehoseBlockStream<C> {}
+
+impl<C: Blockchain> Stream for FirehoseBlockStream<C> {
+    type Item = Result<BlockStreamEvent<C>, Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            match &mut self.state {
+                BlockStreamState::Disconneted => {
+                    info!(
+                        self.ctx.logger,
+                        "Blockstream disconnected, (re-)connecting"; "endpoint uri" => format_args!("{}", self.endpoint),
+                    );
+
+                    let start_block_num: BlockNumber = self
+                        .ctx
+                        .start_blocks
+                        .clone()
+                        .into_iter()
+                        .min()
+                        .unwrap_or(BLOCK_NUMBER_MAX);
```

Comment:
Is this really meant to be `BLOCK_NUMBER_MAX`? It seems it should default to `0`. If that's intended, might be good to comment why `BLOCK_NUMBER_MAX` is used here.